chore(nginx): some minor fixes and documented the config (#3239)
nodejs/build: ansible/www-standalone/resources/config/nodejs.org

    server_tokens off;
    log_not_found off;

>     # We gzip the responses that NGINX does for the following mime-types. We don't use gzip_static as otherwise it would
>     # request a pre-gzip version of each file on the file system, which would hammer the file system with requqests.

Reviewer: ```suggestion
    # We on-the-fly gzip the responses that NGINX does for the following mime-types.
    # We don't use gzip_static as that looks for pre-gzipped files on disk for each request, which we don't have.
```

    gzip on;
    gzip_disable "MSIE [1-6]\.";
    gzip_types text/plain text/css application/javascript text/xml application/xml application/xml+rss image/svg+xml;

>     # These directives are used to prevent our server to continuously attempt to open the same requested files on the file_system
>     # W use this so that file_system accesses get cached during this period of time.
>     # For specific Location directives, such as /dist we also cache the 404's and any other Server Error

Reviewer: ```suggestion
    # These directives prevent our server from continuously attempting to open the same requested files on disk
    # Open file descriptions and basic metadata for each file requested gets cached
    # For specific location blocks, such as /dist, we also cache file not found (404) and other disk errors
```

    open_file_cache max=100000 inactive=300s;
    open_file_cache_valid 120s;
    open_file_cache_min_uses 2;
    open_file_cache_errors off;

    root /home/www/nodejs;
    default_type text/plain;
    index index.html;

>     # We set a default cache of 1 hour and a max age of 4 hours

Reviewer: ```suggestion
    # We set a default cache for browsers of 1 hour, and ask Cloudflare to cache for 4 hours
```

Reviewer 2: Should we set both to 14400? Might be less confusing since CF overrides max-age to 14400.

    add_header Cache-Control "public, max-age=3600, s-maxage=14400";

    error_page 404 /404.html;

>     # For the unencrypted Node.js server, if a request is not done within the /dist folder
>     # We basically rewrite/redirect the request to the https:// schema/protocol

Reviewer: ```suggestion
    # For the unencrypted Node.js server, we redirect any request to the HTTPS site
    # Unless it is within /dist or is for a JSON file
```

    location ~ ^/(?!(dist/|dist$|\.json$)) {
        rewrite ^ https://$host$uri permanent;
    }

>     # This should handle any other possible request done to the unencrypted server
>     # Even tho as seen on the previous Location block for ~ we redirect all non /dist requests to https://

Reviewer: ```suggestion
    # This handles any other possible requests to the unencrypted server
    # With the location block above, and the one below, this should just be JSON files
```

    default_type text/plain;
    index index.html;

    # We set a default cache for browsers of 1 hour, and ask Cloudflare to cache for 4 hours
    add_header Cache-Control "public, max-age=3600, s-maxage=14400";

    error_page 404 /404.html;

    # For the unencrypted Node.js server, we redirect any request to the HTTPS site

>     # Unless it is within /dist or is for a JSON file

Reviewer 3: Would it make sense to comment why we don't redirect for JSON files?

Author: I don't know. That might be a historical reason. cc @richardlau @targos

Reviewer 4: I have no idea.

        }
    }

>     # This Location directive redirects to the list of all Node.js binaries available
>     # of all existing Node.js versions
>     # We use `open_file_cache_errors` that allows the NGINX file cache to also cache that
>     # certain files gave 404's, so that it doesn't keep retrying to open this file that doesn't exist
>     # This policy will still follow the expiry time of the file_cache. It is used mostly to avoid our file system
>     # to be hammered by concurrent requests and mitigate situations where the fs stops fulfiling requests
>     # because of maximum concurrent sockets being open.
>     # We use ^~ to tell NGINX to not process any other Location directive or Rewrite after this match

Reviewer: ```suggestion
    # This location directive exposes the dist directory, including index listings
    # This provides unencrypted access to all Node.js binaries available for all Node.js versions
    # We use `open_file_cache_errors` here to allow NGINX to also cache metadata for missing files on disk
    # This still follows the expiry time of the open_file_cache. It is used mostly to avoid our file system
    # being hammered by concurrent requests and mitigate situations where the fs stops fulfiling requests
    # because of maximum concurrent sockets being open.
    # We use ^~ to tell NGINX to not process any other Location directive or Rewrite after this match
```

            add_header access-control-allow-origin *;
        }
    }

    location = /404.html {

>         # We disable caching of 404 pages as we want to avoid CloudFlare accidentally thinking that a 404 is permanent
>         # many cases 404 pages happen due to the server reaching maximum concurrent file system open() requests
>         # disabling cache allows CloudFlare to re-evaluate the same $uri once our server recovers and then properly cache the page

Reviewer: ```suggestion
        # We disable caching of 404 pages as we always want Cloudflare to check if the file now exists
        # Some 404s may be caused by the server reaching maximum concurrent file system open() requests
        # Disabling cache allows Cloudflare to re-evaluate the same $uri once our server recovers and then properly cache it
```

>         # disabling cache allows CloudFlare to re-evaluate the same $uri once our server recovers and then properly cache the page

Reviewer 2: There's also the case where we publish a new release. Sometimes external automation tries to get the binaries before they are actually published to the server.

Author: Mind adding a suggestion block of what should be added? 👀

    access_log /var/log/nginx/nodejs/nodejs.org-access.log nodejs;
    error_log /var/log/nginx/nodejs/nodejs.org-error.log;

>     # We gzip the responses that NGINX does for the following mime-types. We don't use gzip_static as otherwise it would
>     # request a pre-gzip version of each file on the file system, which would hammer the file system with requqests.

Reviewer: ```suggestion
    # We on-the-fly gzip the responses that NGINX does for the following mime-types
    # We don't use gzip_static as that looks for pre-gzipped files on disk for each request, which we don't have
```
Same as above

    gzip on;
    gzip_disable "MSIE [1-6]\.";
    gzip_types text/plain text/css application/javascript text/xml application/xml application/xml+rss image/svg+xml;

>     # These directives are used to prevent our server to continuously attempt to open the same requested files on the file_system
>     # W use this so that file_system accesses get cached during this period of time.
>     # For specific Location directives, such as /dist we also cache the 404's and any other Server Error

Reviewer: ```suggestion
    # These directives prevent our server from continuously attempting to open the same requested files on disk
    # Open file descriptors and basic metadata for each file requested gets cached
    # For specific location blocks, such as /dist, we also cache file not found (404) and other disk errors
```
Same as above

    open_file_cache max=100000 inactive=300s;
    open_file_cache_valid 120s;
    open_file_cache_min_uses 2;
    open_file_cache_errors off;

    # We set a default language to "en". This is used for the Localized 404 pages
    set $lang en;

    root /home/www/nodejs;
    default_type text/plain;
    index index.html;
    

>     # We set a default cache of 1 hour and a max age of 4 hours

Reviewer: ```suggestion
    # We set a default cache for browsers of 1 hour, and ask Cloudflare to cache for 4 hours
```
Same as above

    add_header Cache-Control "public, max-age=3600, s-maxage=14400";

    # Sets a custom 404 error page for the whole Server region
    # we use the Next.js generated 404 page for all the 404's of our Website
    # including for binaries, assets and et cetera.
    error_page 404 @localized_404;

    # As we don't have a /index.html, we redirect to /en
    location = / {

>         rewrite ^ /en permanent;

Reviewer 2: This rewrite is done with code 302 (temporary) at the CF level

Reviewer 2: I don't know.

Author: I mean, we should probably remove that rewrite from there, as CF would cache this one eitherways.

        }
    }

    # This Location is used for handling static Next.js files. As we don't want to log access
    # to static directories and also we don't want to log not found requests here
    # As this is a static directory that in theory should not change over time, we disable access
    # logs and also cache 404's errors as this folder contents change completely on every 
    # We don't use ^~ as there are other Rewrite directives below that should also be taken into consideration
    # before failing the request with a 404 if it doesn't exist
    location /static {
        access_log off;
        log_not_found off;
        open_file_cache_errors on;
    }

    # This Location directy is used to handle Next.js internal _next directory
    # As this is an internal directory requested by Next.js itself, we disable access
    # logs and also cache 404's errors as this folder contents change completely on every build
    # We use ^~ to tell NGINX to not process any other Location directive or Rewrite after this match
    location ^~ /_next {
        access_log off;
        log_not_found off;
        open_file_cache_errors on;
    }

>     # This Location directive redirects to the list of all Node.js binaries available
>     # of all existing Node.js versions
>     # We use `open_file_cache_errors` that allows the NGINX file cache to also cache that
>     # certain files gave 404's, so that it doesn't keep retrying to open this file that doesn't exist
>     # This policy will still follow the expiry time of the file_cache. It is used mostly to avoid our file system
>     # to be hammered by concurrent requests and mitigate situations where the fs stops fulfiling requests
>     # because of maximum concurrent sockets being open.
>     # We use ^~ to tell NGINX to not process any other Location directive or Rewrite after this match

Reviewer: ```suggestion
    # This location directive exposes the dist directory, including index listings
    # This provides access to all Node.js binaries available for all Node.js versions
    # We use `open_file_cache_errors` here to allow NGINX to also cache metadata for missing files on disk
    # This still follows the expiry time of the open_file_cache. It is used mostly to avoid our file system
    # being hammered by concurrent requests and mitigate situations where the fs stops fulfiling requests
    # because of maximum concurrent sockets being open.
    # We use ^~ to tell NGINX to not process any other location directive or rewrite after this match
```
Same as above

        default_type text/plain;
        open_file_cache_errors on;

        # This rewrite is done to redirect legacy /dist/staging requests to /dist directly
        rewrite ^/dist/staging/(.*)$ /dist/$1 permanent;

        location ~ \.json$ {
            add_header access-control-allow-origin *;
        }
    }

>     # This Location directive redirects to the overall available Downloads on this Server
>     # We use `open_file_cache_errors` that allows the NGINX file cache to also cache that
>     # certain files gave 404's, so that it doesn't keep retrying to open this file that doesn't exist
>     # This policy will still follow the expiry time of the file_cache. It is used mostly to avoid our file system
>     # to be hammered by concurrent requests and mitigate situations where the fs stops fulfiling requests
>     # because of maximum concurrent sockets being open.
>     # We use ^~ to tell NGINX to not process any other Location directive or Rewrite after this match

Reviewer: ```suggestion
    # This location directive exposes the download directory, including index listings
    # We use `open_file_cache_errors` here to allow NGINX to also cache metadata for missing files on disk
    # This still follows the expiry time of the open_file_cache. It is used mostly to avoid our file system
    # being hammered by concurrent requests and mitigate situations where the fs stops fulfiling requests
    # because of maximum concurrent sockets being open.
    # We use ^~ to tell NGINX to not process any other location directive or rewrite after this match
```
Very similar wording to above

        }
    }

    # Redirects the Legacy /documentation endpoint to /api
    # We use ^~ to tell NGINX to not process any other Location directive or Rewrite after this match
    location ^~ /documentation {
        rewrite ^/documentation/api(.*)$ /api$1 redirect;
    }

    # This Location is used for our Metrics Pages
    # We keep `autoindex` on as some folders are autoindex (do not have an index.html)
    # even tho the main /metric endpoint is a build page (statically generated over time)
    # We use ^~ to tell NGINX to not process any other Location directive or Rewrite after this match
    location ^~ /metrics {
        alias /home/dist/metrics;
        autoindex on;
        default_type text/plain;
    }

>     # Store the original language of the request
>     # We'll use this for the 404 in the try_files in @localized_404

Reviewer: ```suggestion
    # When a website 404 occurs, attempt to load the English version of the page
    # if the request was for a localised page.
    # Also, store the original language of the request if it was localised
    # We'll use this language for the 404 in the try_files in @localized_404
```

    location @english_fallback {
        # @TODO: Handle Localization Fallback through Next.js SSR as this is a hacky approach and requires
        # continuous maintenance of the supported languages
        if ($uri ~* ^/(ar|be|ca|de|es|fa|fr|gl|id|it|ja|ka|ko|nl|pt-br|ro|ru|tr|uk|zh-cn|zh-tw)/) {
            set $lang $1;
        }

        rewrite ^/(ar|be|ca|de|es|fa|fr|gl|id|it|ja|ka|ko|nl|pt-br|ro|ru|tr|uk|zh-cn|zh-tw)/(.*)$ /en/$2;
    }

>     # This Location directive is used for handling all 404's requests coming to this Server
>     # by default if the request doesn't come from a /two-letter-code/ localized page (extracted from the @english_fallback Location)
>     # it will then fallback to $lang as being "en" as defined numerous lines above

Reviewer: ```suggestion
    # This location directive handles all 404 responses for the server
    # If the request was a localised website page, use the requested language
    # as set by the @english_fallback location block
    # Otherwise, this will fallback to $lang being "en" as defined numerous lines above
```

    location @localized_404 {

>         # We disable caching of 404 pages as we want to avoid CloudFlare accidentally thinking that a 404 is permanent
>         # many cases 404 pages happen due to the server reaching maximum concurrent file system open() requests
>         # disabling cache allows CloudFlare to re-evaluate the same $uri once our server recovers and then properly cache the page

Reviewer: ```suggestion
        # We disable caching of 404 pages as we always want Cloudflare to check if the file now exists
        # Some 404s may be caused by the server reaching maximum concurrent file system open() requests
        # Disabling cache allows Cloudflare to re-evaluate the same $uri once our server recovers and then properly cache it
```
Same as above

        add_header Cache-Control "private, no-store, max-age=0" always;        

        # If this was a rewritten i18n request from @english_fallback, use the localized 404

>         # if there is no 404 page for that Localized page we will fallback to the English Website 404
>         # as last resort we fallback to NGINX's default 404 on worst scenario. Notice that this result as a [crit] and should never happen

Reviewer: ```suggestion
        # If there is no 404 page for that locale, fallback to the English 404
        # As a last resort, fallback to NGINX's default 404. This should never happen, and will emit a [crit]
```

        try_files /$lang/404.html /en/404.html =404;
    }

    # This Location directive is used to expose our Website GitHub WebHook Logs
    location = /github-webhook.log {
        alias /home/nodejs/github-webhook.log;
        default_type text/plain;
        access_log off;
    }

> 

Reviewer 2: Maybe add a no-store cache header here so we can remove the custom rule at the CF level?

![image](https://user-images.githubusercontent.com/2352663/226096841-91d978ae-ffb1-4602-926f-44e74a6d8510.png)

Author: Done
